src/lib/__tests__/InstantSearch-status.test.ts

```diff
@@ -0,0 +1,213 @@
+    const search = instantsearch({
+      indexName: 'indexName',
+      searchClient: createDelayedSearchClient(20),
+    });
+    search.addWidgets([virtualSearchBox({})]);
+    search.start();
+    await wait(0);
+    expect(search.status).toBe('loading');
+    await wait(20);
+    expect(search.status).toBe('idle');
+
+    search.addWidgets([virtualSearchBox({})]);
+
+    await wait(0);
+
+    expect(search.status).toBe('loading');
+
+    await wait(20);
+
+    expect(search.status).toBe('idle');
+  });
+
+  test('removeWidgets triggers `loading`', async () => {
+    const search = instantsearch({
+      indexName: 'indexName',
+      searchClient: createDelayedSearchClient(20),
+    });
+    const firstWidget = virtualSearchBox({});
+    search.addWidgets([firstWidget, virtualSearchBox({})]);
+    search.start();
+    await wait(0);
+    expect(search.status).toBe('loading');
+    await wait(20);
+    expect(search.status).toBe('idle');
+
+    search.removeWidgets([firstWidget]);
+
+    await wait(0);
+
+    expect(search.status).toBe('loading');
+
+    await wait(20);
+
+    expect(search.status).toBe('idle');
+  });
+
+  test('reaches `stalled` after delay', async () => {
+    const stalledSearchDelay = 400;
+    const requestTime = 500;
+    const search = instantsearch({
+      indexName: 'indexName',
+      stalledSearchDelay,
+      searchClient: createDelayedSearchClient(requestTime),
+    });
+    search.addWidgets([virtualSearchBox({})]);
+    search.start();
+    await wait(0);
+    expect(search.status).toBe('loading');
+
+    await wait(stalledSearchDelay);
+
+    expect(search.status).toBe('stalled');
+
+    await wait(requestTime - stalledSearchDelay);
+
+    expect(search.status).toBe('idle');
+  });
+
+  test('error and status are synchronized', async () => {
+    const search = instantsearch({
+      indexName: 'indexName',
+      searchClient: createSearchClient({
+        search() {
+          return Promise.reject(new Error('Search error'));
+        },
+      }),
+    });
+
+    // prevent rethrow of error
+    search.on('error', () => {});
+
+    search.addWidgets([virtualSearchBox({})]);
+    search.start();
+    await wait(0);
+
+    expect(search.status).toBe('error');
+    expect(search.error).toEqual(expect.any(Error));
+    expect(search.error?.message).toEqual('Search error');
+  });
+
+  test('error can be displayed using `render` event', async () => {
+    expect.assertions(4);
+    const search = instantsearch({
+      indexName: 'indexName',
+      searchClient: createSearchClient({
+        search() {
+          return Promise.reject(new Error('Search error'));
+        },
+      }),
+    });
+
+    // prevent rethrow of error
+    search.on('error', () => {});
+
+    const spy = jest.fn(() => {
+      expect(search.status).toBe('error');
+      expect(search.error).toEqual(expect.any(Error));
+      expect(search.error?.message).toEqual('Search error');
```

Comment:
It could be interesting to actually render something like we would recommend users do.

Instead of testing that the spy was called, we could assert the rendering, which would be closer to actual usage.